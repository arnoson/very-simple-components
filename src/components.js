import { addListeners, getRefs, walkComponent } from './utils/utils.js'

/**
 * @typedef {({ el: HTMLElement, refs: Refs}) => object?} Component
 */

/** @type {Object<string, Component>} */
const components = {}

/**
 * @private
 * @param {HTMLElement} el
 * @returns {Component}
 */
const getComponent = el => components[el.dataset.component]

/**
 * @param {string} name
 * @param {Component} component
 * @returns {Component}
 */
export const registerComponent = (name, component) => {
  components[name] = component
  return component
}

/**
 * Mount a single component.
 * @param {HTMLElement} el
 * @param {object} [component]
 */
export const mountComponent = (el, component = null, isChild = false) => {
  const refs = getRefs(el)
  component = component ?? getComponent(el)
  const context = component({ el, refs })

  // Walk the component and add event listeners.
  walkComponent(el, el => addListeners(el, context))

  // `mountChildComponents` will detect also nested children so we have to call
  // it only once on the top component.
  if (!isChild) {
    mountChildComponents(el)
  }
}

/**
 * Mount all components inside the element.
 * @param {HTMLElement} root - The containing element.
 */
export const mountComponents = (root = document.body) => {
  mountChildComponents(root)
}

/**
 * Mount all components inside an element.
 * @param {HTMLElement} root
 */
const mountChildComponents = root => {
  for (const name in components) {
    const elms = root.querySelectorAll(`[data-component='${name}']`)
    for (let i = 0; i < elms.length; i++) {
      const el = /** @type {HTMLElement} */ (elms[i])
      mountComponent(el, components[name], true)
    }
  }
}
