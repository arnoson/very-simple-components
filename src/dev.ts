import { registerComponent, mountComponents } from '.'

const options = {
  props: {
    number: 42,
    string: 'default',
    bool: false,
    array: () => [],
    obj: () => ({})
  }
}

registerComponent('test', options, ({ el, props }) => {
  console.log(props.bool, el.dataset.bool)
})

mountComponents()
