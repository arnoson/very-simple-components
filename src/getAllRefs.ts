import { walkComponent } from './walkComponent'
import { RefsAll } from './types'

export const getAllRefs = (el: HTMLElement) => {
  const refs: RefsAll = {}
  walkComponent(el, el => {
    const { ref } = el.dataset
    if (ref) {
      refs[ref] ??= []
      refs[ref].push(el)
    }
  })
  return refs
}
