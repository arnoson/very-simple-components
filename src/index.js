import {
  walkComponent,
  parseEvents,
  registerEvents,
  isString
} from './utils/utils.js'

/**
 * @typedef {({ el: HTMLElement, refs: Refs}) => object?} Component
 */

/** @type {Object<string, Component>} */
const components = {}

/**
 * @private
 * @param {HTMLElement} el
 * @returns {Component}
 */
const getComponent = el => components[el.dataset.component]

/**
 * @param {string} name
 * @param {Component} component
 * @returns {Component}
 */
export const registerComponent = (name, component) => {
  components[name] = component
  return component
}

/**
 * Mount a single component.
 * @param {HTMLElement} el
 * @param {object} [component]
 */
export const mountComponent = (el, component = null, isChild = false) => {
  component = component ?? getComponent(el)

  // We first store all events without registering them, because we don't know
  // the refs yet (which we need to create the event handlers' context).
  const allEvents = []
  const refs = {}

  // Search for `data-on-...` event definitions and references.
  // `walkComponent` won't traverse nested components but will execute the
  // callback a last time on the nested component's root element before
  // stopping.
  walkComponent(el, (el, isChildComponent) => {
    // Add a new reference to the list or group multiple references with the
    // same name into an array.
    const { ref } = el.dataset
    if (ref) {
      const entry = refs[ref]
      refs[ref] = entry
        ? Array.isArray(entry)
          ? entry.concat(entry, el)
          : [entry, el]
        : el
    }

    if (!isChildComponent) {
      const events = parseEvents(el)
      events && allEvents.push(events)
    }
  })

  const context = component({ el, refs })
  registerEvents(allEvents, context)

  // Store the context on the el so we can retrieve it with the `very()` helper
  // function.
  /* @ts-ignore */
  el.__very = context

  // `mountChildComponents` will detect also nested children so we have to call
  // it only once on the top component.
  if (!isChild) {
    mountChildComponents(el)
  }
}

/**
 * Mount all components inside the element.
 * @param {HTMLElement} root - The containing element.
 */
export const mountComponents = (root = document.body) => {
  mountChildComponents(root)
}

/**
 * Mount all components inside an element.
 * @param {HTMLElement} root
 */
const mountChildComponents = root => {
  for (const name in components) {
    const elms = root.querySelectorAll(`[data-component='${name}']`)
    for (let i = 0; i < elms.length; i++) {
      const el = /** @type {HTMLElement} */ (elms[i])
      mountComponent(el, components[name], true)
    }
  }
}

/**
 *
 * @param {string | HTMLElement} el An element (or an id).
 */
export const very = el =>
  // @ts-ignore
  (isString(el) ? document.getElementById(el) : el)?.__very
