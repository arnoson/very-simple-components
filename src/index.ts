export { registerComponent } from './registerComponent'
export { mountComponent, mountComponents } from './mountComponent'
export { defineProps } from './props'
