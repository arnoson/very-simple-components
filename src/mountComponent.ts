import { getComponent } from './registerComponent'
import { SimpleElement, SimpleRefs, SimpleRefsAll } from './types'
import { isBuiltInTypeConstructor } from './utils'
import { walkComponent } from './walkComponent'

const parseProp = (value: any, type: string) =>
  type === 'string' ? String(value) : JSON.parse(value)

const stringifyProp = (value: any) =>
  typeof value === 'string' ? value : JSON.stringify(value)

const getDefaultProp = (definition: any) =>
  definition instanceof Function ? definition() : definition

const createPropsProxy = (
  el: HTMLElement,
  definitions: Record<string, any>
) => {
  const get = (_: Object, key: string) => {
    const value = el.dataset[key]
    const definition = definitions[key]
    if (definition === undefined) return value

    const isConstructor = isBuiltInTypeConstructor(definition)
    const providesDefault = !isConstructor
    if (value === undefined && providesDefault)
      return getDefaultProp(definition)

    const type = isConstructor
      ? definition.prototype.constructor.name.toLowerCase()
      : typeof definition

    return parseProp(value, type)
  }

  const set = (_: Object, key: string, value: unknown) => {
    el.dataset[key] = stringifyProp(value)
    return true
  }

  return new Proxy({}, { get, set })
}

const getRefsAll = (el: HTMLElement): SimpleRefsAll<any> => {
  const refs: SimpleRefsAll<any> = {}
  walkComponent(el, el => {
    const { ref } = el.dataset
    if (ref) {
      refs[ref] ??= []
      refs[ref].push(el)
    }
  })
  return refs
}

const mountChildComponents = (el: HTMLElement) => {
  const elements = el.querySelectorAll<HTMLElement>('[data-simple-component]')
  for (let i = 0; i < elements.length; i++) {
    const el = elements[i]
    mountComponent(el, true)
  }
}

/** Mount a single component */
export const mountComponent = (el: HTMLElement, isChild = false) => {
  // Don't re-initialize component.
  if (!(el as any).$component) {
    const refsAll = getRefsAll(el)
    const refs: SimpleRefs = Object.fromEntries(
      Object.entries(refsAll).map(([key, value]) => [key, value[0]])
    )

    const component = getComponent(el)
    if (component) {
      const simpleEl = el as SimpleElement<typeof component>
      const propsDefinitions = component.options.props

      const props = propsDefinitions
        ? createPropsProxy(el, propsDefinitions)
        : el.dataset

      simpleEl.$props = props
      simpleEl.$refs = refs
      simpleEl.$refsAll = refsAll
      simpleEl.$component = component.setup({ el, refs, refsAll, props }) || {}
    }
  }

  if (!isChild) mountChildComponents(el)
}

/** Mount all components inside the element */
export const mountComponents = (root = document.body) => {
  mountChildComponents(root)
}
