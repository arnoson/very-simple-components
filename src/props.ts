type Constructor =
  | NumberConstructor
  | StringConstructor
  | ArrayConstructor
  | ObjectConstructor
  | BooleanConstructor

type Props<T> = {
  [K in keyof T]: T[K] extends Constructor
    ? ReturnType<T[K]> | undefined
    : T[K] extends () => any
    ? ReturnType<T[K]>
    : T[K]
}

const isConstructor = (value: any) =>
  [Number, String, Boolean, Array, Object].includes(value)

const parse = (value: any, type: string) =>
  type === 'string'
    ? String(value)
    : type === 'number'
    ? Number(value)
    : type === 'boolean'
    ? value !== 'false'
    : JSON.parse(value)

const getDefaultValue = (definition: any) =>
  definition instanceof Function ? definition() : definition

export const defineProps = <T>(definitions: T) => {
  return (el: HTMLElement) => {
    const props = Object.fromEntries(
      Object.entries(definitions).map(([key, definition]) => {
        const serializedValue = el.dataset[key]
        const providesDefault = !isConstructor(definition)

        const defaultValue = providesDefault
          ? getDefaultValue(definition)
          : undefined

        const type = providesDefault
          ? typeof defaultValue
          : definition.prototype.constructor.name.toLowerCase()

        const value =
          serializedValue === undefined
            ? defaultValue
            : parse(serializedValue, type)

        return [key, value]
      })
    )

    return props as Props<T>
  }
}
