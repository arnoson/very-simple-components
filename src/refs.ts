import { SimpleRefs, SimpleRefsAll } from './types'
import { walkComponent } from './walkComponent'

export const getRefs = (el: HTMLElement) => {
  const refsAll: SimpleRefsAll = {}
  const addRef = (name: string, el: HTMLElement) => {
    refsAll[name] ??= []
    refsAll[name].push(el)
  }

  walkComponent(el, el => {
    const { ref } = el.dataset
    if (ref) addRef(ref, el)
  })

  const deepRefs = el.querySelectorAll<HTMLElement>('[data-ref*="/"]')
  deepRefs.forEach(refEl => {
    const [parent, name] = refEl.dataset.ref!.split('/')

    const selector = parent.match(/^\((.*)\)$/)?.[1]
    const parentSelector = selector ?? `[data-simple-component='${parent}']`

    const parentEl = el.closest(parentSelector)
    if (parentEl === el) addRef(name, refEl)
  })

  const refs: SimpleRefs = Object.fromEntries(
    Object.entries(refsAll).map(([key, value]) => [key, value[0]])
  )

  return { refs, refsAll }
}
