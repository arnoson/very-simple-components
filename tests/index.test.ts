import { it, expect, vi } from 'vitest'
import { registerComponent, mountComponent, mountComponents } from '../src'

it('mounts a component', () => {
  const component = vi.fn()
  registerComponent('test', component)

  const div = document.createElement('div')
  div.dataset.simpleComponent = 'test'

  mountComponent(div)
  expect(component).toBeCalledWith({ el: div, ref: {}, refs: {} })
})

it('mounts multiple components', () => {
  const componentA = vi.fn()
  registerComponent('a', componentA)

  const componentB = vi.fn()
  registerComponent('b', componentB)

  document.body.innerHTML = `
    <div data-simple-component="a"></div>
    <div data-simple-component="a"></div>
    <div data-simple-component="b"></div>
  `
  mountComponents(document.body)
  expect(componentA).toBeCalledTimes(2)
  expect(componentB).toBeCalledTimes(1)
})

it(`doesn't mount an already mounted component`, () => {
  const component = vi.fn()
  registerComponent('test', component)

  const div = document.createElement('div')
  div.dataset.simpleComponent = 'test'

  mountComponent(div)
  mountComponent(div)

  expect(component).toBeCalledTimes(1)
})

it(`doesn't walk elements with data-simple-ignore attribute`, () => {
  const component = vi.fn()
  registerComponent('test', component)

  document.body.innerHTML = `
    <div data-simple-component="test">
      <div data-simple-ignore data-ref="ref">
        <div data-ref="ref"></div>
      </div>
    </div>
  `
  mountComponents(document.body)
  expect(component).toBeCalledWith(expect.objectContaining({ refs: {} }))
})

it('provides a record of single refs', () => {
  const component = vi.fn()
  registerComponent('test', component)

  document.body.innerHTML = `
    <div data-simple-component="test">
      <div data-ref="myRef"></div>
      <div data-simple-component="another-component">
        <!-- This shouldn't show up in the refs because it belongs to another 
        component -->
        <div data-ref="nestedRef"></div>
      </div>
    </div>
  `
  mountComponents(document.body)
  const myRef = document.querySelector(`[data-ref='myRef']`)
  expect(component).toBeCalledWith(expect.objectContaining({ ref: { myRef } }))
})

it('provides a record of groups of refs with the same name', () => {
  const component = vi.fn()
  registerComponent('test', component)

  document.body.innerHTML = `
    <div data-simple-component="test">
      <div id="ref1" data-ref="myRef"></div>
      <div id="ref2" data-ref="myRef"></div>
      <div id="ref3" data-ref="myRef"></div>
    </div>
  `
  mountComponents(document.body)
  const myRef = [
    document.querySelector('#ref1'),
    document.querySelector('#ref2'),
    document.querySelector('#ref3')
  ]
  expect(component).toBeCalledWith(expect.objectContaining({ refs: { myRef } }))
})

it(`exposes the component function's return value`, () => {
  const exposed = { test: () => {} }
  registerComponent('test', () => exposed)

  document.body.innerHTML = `
    <div data-simple-component="test" id="my-id"></div>
  `
  mountComponents(document.body)
  expect(document.getElementById('my-id').$component).toBe(exposed)
})
