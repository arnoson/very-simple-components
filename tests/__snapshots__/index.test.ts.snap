// Vitest Snapshot v1

exports[`interferes prop types from default values 1`] = `
{
  "array": [
    1,
    2,
    3,
  ],
  "bool": true,
  "number": 1,
  "obj": {
    "hello": "world",
  },
  "string": "text",
}
`;

exports[`parses props 1`] = `
{
  "array": [
    1,
    2,
    3,
  ],
  "bool": true,
  "number": 1,
  "obj": {
    "hello": "world",
  },
  "string": "text",
}
`;

exports[`provides default values for props 1`] = `
{
  "array": [],
  "bool": true,
  "number": 10,
  "obj": {},
  "string": "hello",
}
`;
